core/include/detray/utils/indexing.hpp

```diff
@@ -17,4 +17,100 @@ dindex constexpr dindex_invalid = std::numeric_limits<dindex>::max();
 using dindex_range = darray<dindex, 2>;
 using dindex_sequence = dvector<dindex>;
 
+/** Small type to tie an object type and an index into a container together.
+ *
+ * @tparam id_type Represents the indexed type
+ * @tparam intex_type The type of indexing needed for the indexed types
+ * container
+ */
+template <typename id_type = unsigned int, typename index_type = dindex>
+struct typed_index {
+    id_type _object_id;
+    index_type _index;
+
+    /** Equality operator */
+    DETRAY_HOST_DEVICE
+    bool operator==(const typed_index<id_type, index_type>& rhs) const {
+        return (_object_id == rhs._object_id && _index == rhs._index);
+    }
+
+    /** Arithmetic operators*/
+    DETRAY_HOST_DEVICE
+    typed_index<id_type, index_type> operator+(
+        const typed_index<id_type, index_type>& rhs) const {
+        return {_object_id, _index + rhs._index};
+    }
+
+    DETRAY_HOST_DEVICE
+    typed_index<id_type, index_type> operator+(const index_type& index) const {
+        return {_object_id, _index + index};
+    }
+
+    DETRAY_HOST_DEVICE
+    typed_index<id_type, index_type> operator-(
+        const typed_index<id_type, index_type>& rhs) const {
+        return {_object_id, _index - rhs._index};
+    }
+
+    DETRAY_HOST_DEVICE
+    typed_index<id_type, index_type> operator-(const index_type& index) const {
+        return {_object_id, _index - index};
+    }
+
+    DETRAY_HOST_DEVICE
+    typed_index<id_type, index_type>& operator+=(
+        const typed_index<id_type, index_type>& rhs) {
+        _index += rhs._index;
+        return *this;
+    }
+
+    DETRAY_HOST_DEVICE
+    typed_index<id_type, index_type>& operator+=(const index_type& index) {
+        _index += index;
+        return *this;
+    }
+
+    DETRAY_HOST_DEVICE
+    typed_index<id_type, index_type>& operator-=(
+        const typed_index<id_type, index_type>& rhs) {
+        _index -= rhs._index;
+        return *this;
+    }
+
+    DETRAY_HOST_DEVICE
+    typed_index<id_type, index_type>& operator-=(const index_type& index) {
+        _index -= index;
+        return *this;
+    }
+
+    /** Only make the prefix operator available */
+    DETRAY_HOST_DEVICE
+    typed_index<id_type, index_type>& operator++() {
+        ++_index;
+        return *this;
+    }
+};
+
+namespace detail {
+
+using std::get;
```

Comment:
Do we need it?

Reply:
You may be right. If we want to enforce the use of ids instead of plain integers, we should be thorough and remove it